### ISSUE TITLE
zrepl crashing on arm64 due to invalid uzfs zc_nvlist_dst_size handling

### ISSUE THREAD

**Issue by sgielen** [openebs/openebs#3346]
## Description
I am seeing crashes in zrepl that seem to boil down to invalid zc_nvlist_dst_size handling in some cases.

The code probably exhibits the same undefined behaviour on amd64, but it may not lead to crashes there.

In `uzfs_ioctl_init`, buffers are allocated to contain nvlist items:

- `zc->zc_nvlist_src` is set to `malloc(zc->zc_nvlist_src_size)` (i.e. the buffer is `zc_nvlist_src_size` **bytes** large)
- `zc->zc_nvlist_dst` is set to `malloc(zc->zc_nvlist_dst_size)`
- `zc->zc_nvlist_conf` is set to `malloc(zc->zc_nvlist_conf_size)`

I am looking at an invocation where zrepl gets a `ZFS_IOC_ERROR_LOG` from the cstor pool mgmt process. The `zc_nvlist_dst_size` is set to 17 when `uzfs_ioctl_init` is called, so `zc_nvlist_dst` is initialized to a buffer of 17 **bytes**. However, later in the code, I see `zc_nvlist_dst_size` refers not to the amount of bytes, but to the amount of **objects**.

In one run of the code, annotated with additional output, I see that `uzfs_ioctl_init` initializes `zc_nvlist_dst=0x907d430` based on a size of `17`. I.e. `0x907d430` is a buffer of 17 bytes.

While handling the `ZFS_IOC_ERROR_LOG` ioctl, `zfs_ioc_error_log` is called with the same values set: `zc_nvlist_dst=0x907d430` and `zc_nvlist_dst_size=17`. It calls into `spa_get_errlog(spa, zc->zc_nvlist_dst, &count)` where `uaddr=0x907d430` and `count=17`. This function calls into the static function `process_error_log(..., uaddr, &count)`, and that function assumes that the given address can contain `count` blocks of size `sizeof(zbookmark_phys_t)`. Since `count=17` and `addr=zc->zc_nvlist_dst=0x907d430` at this point, and it tries to write to `addr + (*count - 1) * sizeof(zbookmark_phys_t)`, it writes to `0x907d590` which is 0x160 (352) bytes beyond the start of the 17-byte buffer.

I'm not sure if `zc_nvlist_dst_size` should refer to the amount of **objects** and the ioctl or the allocation is incorrect, or if it should refer to the amount of **bytes** and the usage of the nvlist is incorrect. I do see, in other places, that `zc_nvlist_dst_size` is usually much larger, e.g. I see invocations where it has `zc_nvlist_dst_size=16384` and even `65536`. I do know that this particular invocation of the `ZFS_IOC_ERROR_LOG` ioctl leads to a crash, and I think the root cause might be confusion whether it refers to objects or sizes.

**Comment by sgielen** [comment]
I found this by running `zrepl` insize valgrind, which pretty quickly finds this invalid write (after silencing lots and lots of warnings of uninitialized memory by initializing all allocations to zero, just like I did for #3028).

If I let `uzfs_ioctl_init` always malloc `zc_nvlist_dist_size * 32` bytes, it may allocate too much in some cases, but the valgrind warning goes away, showing that this is indeed the cause.

**Comment by sgielen** [comment]
Another relevant data point is in `zpool_get_errlog` in `lib/libzfs/libzfs_pool.c`, which performs the ioctl on the client side.

```
        zc.zc_nvlist_dst = (uintptr_t)zfs_alloc(zhp->zpool_hdl,
            count * sizeof (zbookmark_phys_t));
        zc.zc_nvlist_dst_size = count;
```

Here, `zc_nvlist_dst_size` is the amount of **objects**, just before the ioctl is sent (which, in cStor, means a UNIX domain socket message). However, on the server side, `dst_size` is always interpreted as the number of **bytes**, causing the issue.

**Comment by pawanpraka1** [comment]
@sgielen this is great finding. ZFS_IOC_ERROR_LOG ioctl uses dst_size as count of (zbookmark_phys_t) object it can put. We should probably consider this while doing uzfs_ioctl_init.

**Comment by sgielen** [comment]
@pawanpraka1 are you aware of any other ioctls where the `dst_size` (or `src_size` or `conf_size`) don't refer to bytes? If you'd like I can submit a PR so that uzfs_ioctl_init takes this exception into account while allocating the buffers?

**Comment by pawanpraka1** [comment]
@sgielen I searched and could not find any other IOCTL with such a logic. Please go ahead and raise the PR to fix this issue.

**Comment by sgielen** [comment]
@pawanpraka1 Please review openebs/libcstor#82. :-)

**Comment by sgielen** [comment]
Once my cluster is repaired I'll update to the new `ci` builds including openebs/libcstor#82, to confirm the issue is fixed. Then I'll close this :)

**Comment by sgielen** [comment]
I haven't forgotten, by the way, the cluster is still repairing... :-)

**Comment by sgielen** [comment]
(Still repairing, still intend to confirm the issue is resolved in 2.7...)

**Comment by pawanpraka1** [comment]
@sgielen what is the status of this issue? Can we close this one?

**Comment by sgielen** [comment]
I'm still waiting for the cluster rebuild to finish. Unfortunately, I can't see how far it is, only that one CVR is still in Rebuilding and its Used/Allocated is sometimes growing by one GB. Perhaps it would be best to just close this one, and I'll reopen if the issue comes back after I upgrade to the 2.7 release.

**Comment by sgielen** [comment]
For historical purposes: eventually, it turned out that the repair was failing at the same point, causing it to never succeed, hence taking forever. This was investigated and fixed as a part of #3388.

Since then, the cluster has repaired and I have upgraded to OpenEBS 2.8.0, and can confirm that the issue is indeed resolved in that release.
